utils/src/utils.cpp

```diff
@@ -1,44 +1,78 @@
 #include "utils.hpp"
 
+#include <chrono>
 #include <vector>
 
 namespace util {
 
+uint64_t get_time_ms() {
+    const auto timestamp = std::chrono::system_clock::now();
+    return std::chrono::duration_cast<std::chrono::milliseconds>(
+               timestamp.time_since_epoch())
+        .count();
+}
+
 constexpr uint8_t hex_to_nibble(const char & ch)
 {
-  return ( ch >= '0' && ch <= '9') ? ch - 48 : ((ch >= 'A' && ch <= 'F' ) ? ch - 55 : ((ch >= 'a' && ch <= 'f' ) ? ch - 87 : 0));
+    return ( ch >= '0' && ch <= '9') ? ch - 48 : ((ch >= 'A' && ch <= 'F' ) ? ch - 55 : ((ch >= 'a' && ch <= 'f' ) ? ch - 87 : 0));
 }
 
 constexpr uint8_t hexpair_to_byte(const char & hi, const char & lo)
 {
-  return hex_to_nibble(hi) << 4 | hex_to_nibble(lo);
+    return hex_to_nibble(hi) << 4 | hex_to_nibble(lo);
 }
 
 std::string hex64_to_base32z(const std::string &src)
 {
-  // decode to binary
-  std::vector<uint8_t> bin;
-  // odd sized is invalid
-  if(src.size() & 1)
-    return "";
-  {
-    auto itr = src.begin();
-    while(itr != src.end())
+    // decode to binary
+    std::vector<uint8_t> bin;
+    // odd sized is invalid
+    if(src.size() & 1)
+        return "";
     {
-      const char hi = *itr;
-      ++itr;
-      const char lo = *itr;
-      ++itr;
-      bin.emplace_back(hexpair_to_byte(hi,lo));
+        auto itr = src.begin();
+        while(itr != src.end())
+        {
+            const char hi = *itr;
+            ++itr;
+            const char lo = *itr;
+            ++itr;
+            bin.emplace_back(hexpair_to_byte(hi,lo));
+        }
+    }
+    // encode to base32z
+    char buf[64] = {0};
+    std::string result;
+    if (char const *dest = base32z_encode(bin, buf))
+        result = dest;
+
+    return result;
+}
+
+bool parseTimestamp(const std::string& timestampString, const uint64_t& ttl,
+                    uint64_t& timestamp) {
+    try {
+        timestamp = std::stoull(timestampString);
+    } catch (...) {
+        return false;
     }
-  }
-  // encode to base32z
-  char buf[64] = {0};
-  std::string result;
-  if (char const *dest = base32z_encode(bin, buf))
-    result = dest;
-
-  return result;
+    const uint64_t cur_time = get_time_ms();
+
+    // Prevent overflow
+    if (cur_time >= UINT64_MAX - ttl)
```

Comment:
DOOT DOOT
`cur_time` is in milliseconds while ttl is in seconds...
But regardless, you can probably drop the whole overflow check:
- cur_time is the unix timestamp * 1000
- ttl is at most `96 * 60 * 60` = 345600

So if you calculate `(UINT64_MAX - 345600)/1000` you get the current time at which it will overflow:
> 04/03/584556019 @ 2:20pm (UTC)

And that will be a Wednesday, in case you were wondering 🤣

Reply:
This check was more to prevent malicious clients causing undefined behaviour (to be fair idk what it could possibly do but better safe than sorry?)

Reply:
I can't think of any problem with the overflow in this case, but yeah, we could still check for that.

Reply:
Actually whoops I was doing this calculation wrong anyway, as of the last PR exp_time should be calculated using *timestamp* not cur_time
Fixing